README
POE-Component-Client-Stomp version 0.11
=======================================

This module is a object wrapper around Stomp client. It provides
support for programs that want to communicate with a Stomp server such as
POE::Component::Server::MessageQueue or ActiveMQ from the Apache Foundation.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  POE
  POE::Filter::Stomp
  POE::Component::Client::Stomp::Utils

COPYRIGHT AND LICENCE

Copyright (C) 2007 by Kevin L. Esteb

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension POE::Component::Client::Stomp.

0.01  Thu Aug 30 13:48:39 2007
    - original version; created by h2xs 1.23 with options
        -Axc -n POE::Component::Client::Stomp

0.02  03-Oct-2007 K.Esteb
    - Changed the module to be object oriented. I was having problems with
      sub classing the module to reduce complexity. I also wanted a dispatcher
      based on message types. This version does those things.

0.03  09-Oct-2007 K.Esteb
    - Changed event/method "handle_send" to "send_data", it seemed more
        syntactically correct with the new "gather_data" event/method.
      Added successors for stomp, config, host and port.
      Added a log method, this method currently knows "warn" and "error"
        levels. This should be overridden by your class, for your preferred
        logger.
      Added a event/method "gather_data", it was needed to send the
        object to the method.
      Changed the internals to use the new assessors.
      Added a couple of examples.
      Made the reconnection code more robust. Probably needs more work.

0.04  15-Oct-2007 K.Esteb
      Fixed the "Broken pipe" error.
      Figured out how to require other modules. So Net::Stomp,
        POE::Filter::Stomp and POE::Component::Client::Stomp::Utils are
        now required modules.
      Added a method "handle_shutdown" this should be overridden if specific 
        stuff needs to be done when a shutdown has been initiated. By default 
        nothing will be done.
      Fixed a syntax mistake in the documentation.
      Added a granulated scheme for reconnection attempts.

0.05  17-Dec-2007 K.Esteb
      Fixed the "connection timeout" error.
      Fixed the documentation so it would format correctly.
      Added an event/method "connection_down" to signal when the active 
        connection is down. This needs to be overridden. By default nothing
        will be done.
      Added tests to make sure everything works correctly. These tests will
        fail if there is no MQ listening on port 61613 at localhost. Open
        to suggestions on how to fix this.
      Fixed Makefile.PL

0.06  16-Apr-2008 K.Esteb
      Bug fix release for RT #34743, #34742.
      POE is now a required module. Seems rather obvious, but I had some bug
        reports on this one.
      Removed 02_basic.t, Received a bug report on it not working. Sigh, 
        release 0.05 explains why it may not work. So I guess this "fixes" 
        the above problem.

0.07  23-Jun-2008 K.Esteb
      Alejandro Imass made the suggestion that reconnection's should recycle 
        after they have expired. This version includes his logic to do so.

0.08  15-Dec-2008 K.Esteb
      Added some new error codes for reconnection attempts.

0.09  24-Jul-2009 K.Esteb
      Added some more error codes for reconnection attempts.

0.10  04-Nov-2009 K.Esteb
      Added some more error codes for reconnection attempts.
      Changed the logic from 0.07 for retrying reconnection's. It didn't seem
        to work correctly.

0.11  19-Aug-2010 K.Esteb
      Added a "EnableKeepAlive" config option to set SO_KEEPALIVE at the
        socket level. 
      Also added code to set the socket for reasonable levels of 
        keepalive packet generation. Unfortunately the constants; 
        TCP_KEEPCNT, TCP_KEEPIDLE and TCP_KEEPINTVL have different values 
        depending on the OS being used. As such, the this feature only 
        works on Linux and AIX. 



examples/dump_queue.pl
#!/usr/bin/perl
#
# File: dump_queue.pl
# Date: 27-Sep-2007
# By  : Kevin Esteb
#
# Simple test program to test POE interaction.
#

package Client;

use POE;
use Data::Dumper;
use base qw(POE::Component::Client::Stomp);

use strict;
use warnings;

# ----------------------------------------------------------------------

sub handle_connection {
    my ($kernel, $self) = @_[KERNEL,OBJECT];

    my $frame;
    my $buffer = sprintf("Connected to %s on %s", $self->host, $self->port);

    $self->log($kernel, 'info', $buffer);

    $frame = $self->stomp->connect({login => 'testing', 
                                    passcode => 'testing'});
    $kernel->yield('send_data' => $frame);

}

sub handle_connected {
    my ($kernel, $self, $frame) = @_[KERNEL,OBJECT,ARG0];

    my $nframe;

    $nframe = $self->stomp->subscribe({destination => $self->config('Queue'), 
                                       ack => 'client'});
    $kernel->yield('send_data' => $nframe);

}

sub handle_message {
    my ($kernel, $self, $frame) = @_[KERNEL,OBJECT,ARG0];

    my $message_id = $frame->headers->{'message-id'};
    my $nframe = $self->stomp->ack({'message-id' => $message_id});
    my $buffer = sprintf("Received message #%s", $message_id);
    $self->log($kernel, 'info', $buffer);
    print Dumper($frame) if ($self->config('Dump'));
    $kernel->yield('send_data' => $nframe);

}

sub log {
    my ($self, $kernel, $level, @args) = @_;

    if ($level eq 'error') {

        print "Error: @args\n";

    } elsif ($level eq 'warn') {

        print "Warn : @args\n";

    } elsif ($level eq 'debug') {

        print "Debug: @args\n" if $self->config('Debug');

    } else {

        print "Info : @args\n";

    }

}

sub handle_shutdown {
    my ($self, $kernel) = @_;

    print "Shutting down\n";

}

# =====================================================================

package main;

use POE;
use Getopt::Long;

use strict;
use warnings;

my $dump = 0;
my $debug = 0;
my $port = '61613';
my $hostname = 'localhost';
my $queue = '/queue/testing';

my $VERSION = '0.01';

# ----------------------------------------------------------------------

sub handle_signals {

    $poe_kernel->yield('shutdown');

}

sub usage {

    my ($Script) = ( $0 =~ m#([^\\/]+)$# );
    my $Line = "-" x length( $Script );

    print << "EOT";
$Script
$Line
dump_queue - Dump a STOMP message queue.
Version: $VERSION

Usage:

    $0 [--hostname] <hostname>
    $0 [--port] <port number>
    $0 [--queue] <queue name>
    $0 [--dump]
    $0 [--help]
    $0 [--debug]

    --hostname..The host where the server is localed
    --port......The port to connect too
    --queue.....The message queue to listent too
    --dump......A flag to indicate dumping of the message body
    --debug.....Print debugging messages
    --help......Print this help message.

Examples:

    $0 --hostname mq.example.com --port 61613 --queue /queue/testing
    $0 --help

EOT

}

sub setup {

    my $help;

    GetOptions('help|h|?' => \$help, 
               'hostname=s' => \$hostname,
               'port=s' => \$port,
               'queue=s' => \$queue,
               'dump' => \$dump,
               'debug' => \$debug);

    if ($help) {

        usage();
        exit 0;

    }

}

main: {

    setup();

    Client->spawn(
        RemoteAddress => $hostname,
        RemotePort => $port,
        Alias => 'testing',
        Queue => $queue,
        Dump => $dump,
        Debug => $debug,
    );

    $poe_kernel->state('got_signal', \&handle_signals);
    $poe_kernel->sig(INT => 'got_signal');

    $poe_kernel->run();

    exit 0;

}



examples/feed_queue.pl
#!/usr/bin/perl
# 
# File: feed_queue.pl
# Date: 27-Sep-2007
# By  : Kevin Esteb
#
# send a message to the queue 'foo'
#

use Net::Stomp;
use Getopt::Long;

use strict;
use warnings;

# ----------------------------------------------------------------------
# Global variables
# ----------------------------------------------------------------------

my $stomp;
my $count = 10;
my $port = '61613';
my $hostname = 'localhost';
my $queue = '/queue/testing';
my $VERSION = '0.01';

# ----------------------------------------------------------------------

sub usage {

    my ($Script) = ( $0 =~ m#([^\\/]+)$# );
    my $Line = "-" x length( $Script );

    print << "EOT";
$Script
$Line
feed_queue - Feed a STOMP message queue.
Version: $VERSION

Usage:

    $0 [--hostname] <hostname>
    $0 [--port] <port number>
    $0 [--queue] <queue name>
    $0 [--count] <number>
    $0 [--help]

    --hostname..The host where the server is localed
    --port......The port to connect too
    --queue.....The message queue to listent too
    --count.....The number of times to send the message
    --help......Print this help message.

  Examples:

    $0 --hostname mq.example.com --port 61613 --queue /queue/testing
    $0 --help

EOT

}

sub setup {

    my $help;

    GetOptions('help|h|?' => \$help, 
               'hostname=s' => \$hostname,
               'port=s' => \$port,
               'queue=s' => \$queue,
               'count=i' => \$count);

    if ($help) {

        usage();
        exit;

    }

}

main: {

    setup();

    $stomp = Net::Stomp->new({ hostname => $hostname, port => $port });
    $stomp->connect( { login => 'testing', passcode => 'testing' } );

    for (my $x = 0; $x < $count; $x++) {

        $stomp->send({ destination => $queue, body => "test message: $x"  });

    }

    $stomp->disconnect;

}



examples/stress_queue.pl
#!/usr/bin/perl
#
# File: stress_queue.pl
# Date: 15-Oct-2007
# By  : Kevin Esteb
#
# Simple test program to test POE interaction.
#

package Client;

use POE;
use Data::Dumper;
use base qw(POE::Component::Client::Stomp);

use strict;
use warnings;

# ----------------------------------------------------------------------

sub spawn {
    my $class = shift;

    my %args = @_;
    my $self = $class->SUPER::spawn(%args);

    $self->{counter} = 0;
    $self->{alarm_id} = 0;

    return $self;

}

sub handle_connection {
    my ($kernel, $self) = @_[KERNEL,OBJECT];

    my $frame;
    my $buffer = sprintf("Connected to %s on %s", $self->host, $self->port);

    $self->log($kernel, 'info', $buffer);
    $frame = $self->stomp->connect({login => 'testing', 
                                    passcode => 'testing'});
    $kernel->yield('send_data' => $frame);

}

sub handle_connected {
    my ($kernel, $self, $frame) = @_[KERNEL,OBJECT,ARG0];

    $kernel->yield('gather_data');

}

sub gather_data {
    my ($kernel, $self) = @_[KERNEL,OBJECT];
    
    $self->{counter}++;
    my $body = sprintf("Message #%s", $self->{counter});
    my $frame = $self->stomp->send({destination => $self->config('Queue'),
                                    data => $body});

    $self->{alarm_id} = $kernel->delay_set('gather_data', 15);
    $kernel->yield('send_data' => $frame);

}

sub log {
    my ($self, $kernel, $level, @args) = @_;

    if ($level eq 'error') {

        print "Error: @args\n";

    } elsif ($level eq 'warn') {

        print "Warn : @args\n";

    } elsif ($level eq 'debug') {

        print "Debug: @args\n" if $self->config('Debug');

    } else {

        print "Info : @args\n";

    }

}

sub handle_shutdown {
    my ($self, $kernel) = @_;

    $kernel->alarm_remove($self->{alarm_id});
    print "Shutting down\n";

}

# =====================================================================

package main;

use POE;
use Getopt::Long;

use strict;
use warnings;

my $debug = 0;
my $port = '61613';
my $hostname = 'localhost';
my $queue = '/queue/testing';

my $VERSION = '0.01';

# ----------------------------------------------------------------------

sub handle_signals {

    $poe_kernel->yield('shutdown');

}

sub usage {

    my ($Script) = ( $0 =~ m#([^\\/]+)$# );
    my $Line = "-" x length( $Script );

    print << "EOT";
$Script
$Line
stress_queue - Send messages to a STOMP message queue.
Version: $VERSION

Usage:

    $0 [--hostname] <hostname>
    $0 [--port] <port number>
    $0 [--queue] <queue name>
    $0 [--dump]
    $0 [--help]
    $0 [--debug]

    --hostname..The host where the server is localed
    --port......The port to connect too
    --queue.....The message queue to listent too
    --debug.....Print debugging messages
    --help......Print this help message.

Examples:

    $0 --hostname mq.example.com --port 61613 --queue /queue/testing
    $0 --help

EOT

}

sub setup {

    my $help;

    GetOptions('help|h|?' => \$help, 
               'hostname=s' => \$hostname,
               'port=s' => \$port,
               'queue=s' => \$queue,
               'debug' => \$debug);

    if ($help) {

        usage();
        exit 0;

    }

}

main: {

    setup();

    Client->spawn(
        RemoteAddress => $hostname,
        RemotePort => $port,
        Alias => 'testing',
        Queue => $queue,
        Debug => $debug,
    );

    $poe_kernel->state('got_signal', \&handle_signals);
    $poe_kernel->sig(INT => 'got_signal');

    $poe_kernel->run();

    exit 0;

}



lib/POE/Component/Client/Stomp.pm
package POE::Component::Client::Stomp;

use POE;
use Carp;
use Socket ':all'; 
use POE::Filter::Stomp;
use POE::Wheel::ReadWrite;
use POE::Wheel::SocketFactory;
use POE::Component::Client::Stomp::Utils;

use 5.8.2;
use strict;
use warnings;

my $TCP_KEEPCNT = 0;
my $TCP_KEEPIDLE = 0;
my $TCP_KEEPINTVL = 0;

use constant DEFAULT_HOST  => 'localhost';
use constant DEFAULT_PORT  => 61613;

if ($^O eq "aix") {           # from /usr/include/netinet/tcp.h

    $TCP_KEEPIDLE  = 0x11;
    $TCP_KEEPINTVL = 0x12;
    $TCP_KEEPCNT   = 0x13;

} elsif ($^O eq "linux"){     # from /usr/include/netinet/tcp.h

    $TCP_KEEPIDLE  = 4;
    $TCP_KEEPINTVL = 5;
    $TCP_KEEPCNT   = 6;

}

my @errors = qw(0 32 68 73 78 79 110 104 111);
my @reconnections = qw(60 120 240 480 960 1920 3840);

our $VERSION = '0.11';

# ---------------------------------------------------------------------

sub spawn {
    my $package = shift;

    croak "$package requires an even number of parameters" if @_ & 1;

    my %args = @_;
    my $self = bless ({}, $package);

    $args{Alias} = 'stomp-client' unless defined $args{Alias} and $args{Alias};
    $args{RetryReconnect} = 1 unless defined $args{RetryReconnect};
    $args{EnableKeepAlive} = defined($args{EnableKeepAlive}) ? $args{EnableKeepAlive} : 0;

    $self->{CONFIG} = \%args;
    $self->{count} = scalar(@reconnections);
    $self->{stomp} = POE::Component::Client::Stomp::Utils->new();
    $self->{attempts} = 0;

    POE::Session->create(
        object_states => [
            $self => { 
                _start => '_client_start',
                _stop => '_client_stop',
                shutdown => '_client_close',
                reconnect => '_server_connect',
            },
            $self => [ qw( _server_connect _server_connected _server_connection_failed 
                           _server_error _server_message _client_close
                           handle_message handle_receipt handle_error
                           handle_connected handle_connection
                           send_data gather_data connection_down) ],
        ],
        (ref $args{options} eq 'HASH' ? (options => $args{options}) : () ),
    );

    return $self;

}

sub _client_start {
    my ($kernel, $self) = @_[KERNEL, OBJECT];

    $self->log($kernel, 'debug', '_client_start()');

    $kernel->alias_set($self->config('Alias'));
    $kernel->yield('_server_connect');

}

sub _server_connect {
    my ($kernel, $self) = @_[KERNEL, OBJECT];

    $self->log($kernel, 'debug', '_server_connect()');

    $self->{Listner} = POE::Wheel::SocketFactory->new(
        RemoteAddress  => $self->config('RemoteAddress') || DEFAULT_HOST,
        RemotePort     => $self->config('RemotePort') || DEFAULT_PORT,
        SocketType     => SOCK_STREAM,
        SocketDomain   => AF_INET,
        Reuse          => 'no',
        SocketProtocol => 'tcp',
        SuccessEvent   => '_server_connected',
        FailureEvent   => '_server_connection_failed',
    );

}

sub _client_stop {
    my ($kernel, $self) = @_[KERNEL, OBJECT];

    $self->log($kernel, 'debug', '_client_stop()');

}

sub _client_close {
    my ($kernel, $self) = @_[KERNEL, OBJECT];

    $self->log($kernel, 'debug', '_client_close()');

    delete $self->{Listner};
    delete $self->{Wheel};
    $self->handle_shutdown($kernel);
    $kernel->alias_remove($self->config('Alias'));

}

sub _server_connected {
    my ($kernel, $self, $socket, $peeraddr, $peerport, $wheel_id) = 
       @_[KERNEL, OBJECT, ARG0 .. ARG3];

    my $os = $^O;

    $self->log($kernel, 'debug', '_server_connected()');

    my $wheel = POE::Wheel::ReadWrite->new(
        Handle => $socket,
        Filter => POE::Filter::Stomp->new(),
        InputEvent => '_server_message',
        ErrorEvent => '_server_error',
    );

    if ($self->config('EnableKeepAlive')) {

        $self->log($kernel, 'debug', 'keepalive activated');

        # turn keepalive on, this should send a keepalive 
        # packet once every 2 hours according to the RFC.

        setsockopt($socket, SOL_SOCKET,  SO_KEEPALIVE,  1);

        if (($os eq 'linux') or ($os eq 'aix')) {

            $self->log($kernel, 'debug', 'adjusting keepalive activity');

            # adjust from system defaults, all values are in seconds.
            # so this does the following:
            #     every 15 minutes send upto 3 packets at 5 second intervals
            #         if no reply, the connection is down.

            setsockopt($socket, IPPROTO_TCP, $TCP_KEEPIDLE,  900);  # 15 minutes
            setsockopt($socket, IPPROTO_TCP, $TCP_KEEPINTVL, 5);    # 
            setsockopt($socket, IPPROTO_TCP, $TCP_KEEPCNT,   3);    # 

        }

    }

    my $host = gethostbyaddr($peeraddr, AF_INET);

    $self->{attempts} = 0;
    $self->{Wheel} = $wheel;
    $self->{Host} = $host;
    $self->{Port} = $peerport;

    $kernel->yield('handle_connection');

}

sub _server_connection_failed {
    my ($kernel, $self, $operation, $errnum, $errstr, $wheel_id) = 
        @_[KERNEL, OBJECT, ARG0 .. ARG3];

    $self->log($kernel, 'debug', '_server_connection_failed()');
    $self->log($kernel, 'error', "Operation: $operation; Reason: $errnum - $errstr");

    delete $self->{Listner};
    delete $self->{Wheel};

    foreach my $error (@errors) {

        $self->_reconnect($kernel) if ($errnum == $error);

    }

}

sub _server_error {
    my ($kernel, $self, $operation, $errnum, $errstr, $wheel_id) = 
        @_[KERNEL, OBJECT, ARG0 .. ARG3];

    $self->log($kernel, 'debug', '_server_error()');
    $self->log($kernel, 'error', "Operation: $operation; Reason: $errnum - $errstr");

    delete $self->{Listner};
    delete $self->{Wheel};

    $kernel->yield('connection_down');

    foreach my $error (@errors) {

        $self->_reconnect($kernel) if ($errnum == $error);

    }

}

sub _server_message {
    my ($kernel, $self, $frame, $wheel_id) = @_[KERNEL, OBJECT, ARG0, ARG1];

    $self->log($kernel, 'debug' , '_server_message()');

    if ($frame->command eq 'CONNECTED') {

        $self->log($kernel, 'debug' , "received a \"CONNECTED\" message");
        $kernel->yield('handle_connected', $frame);

    } elsif ($frame->command eq 'MESSAGE') {

        $self->log($kernel, 'debug' , "received a \"MESSAGE\" message");
        $kernel->yield('handle_message', $frame);

    } elsif ($frame->command eq 'RECEIPT') {

        $self->log($kernel, 'debug' , "received a \"RECEIPT\" message");
        $kernel->yield('handle_receipt', $frame);

    } elsif ($frame->command eq 'ERROR') {

        $self->log($kernel, 'debug' , "received an \"ERROR\" message");
        $kernel->yield('handle_error', $frame);

    } else {
        
        $self->log($kernel, 'warn', "Unknown message type: $frame->command");

    }

}

sub _reconnect {
    my ($self, $kernel) = @_;

    my $retry;

    $self->log($kernel, 'debug', "Attempts: $self->{attempts}, Count: $self->{count}");

    if ($self->{attempts} < $self->{count}) {

        my $delay = $reconnections[$self->{attempts}];
        $self->log($kernel, 'warn', "Attempting reconnection: $self->{attempts}, waiting: $delay seconds");
        $self->{attempts}++;
        $kernel->delay('reconnect', $delay);

    } else {

        $retry = $self->config('RetryReconnect') || 0;

        if ($retry) {

            $self->log($kernel, 'warn', 'Cycling reconnection attempts, but not shutting down...');
            $self->{attempts} = 0;
            $kernel->yield('reconnect');

        } else {

            $self->log($kernel, 'warn', 'Shutting down, to many reconnection attempts');
            $kernel->yield('shutdown'); 

        }

    }

}

# ---------------------------------------------------------------------
# Public accessors
# ---------------------------------------------------------------------

sub stomp {
    my $self = shift;

    return $self->{stomp};

}

sub config {
    my ($self, $arg) = @_;

    return $self->{CONFIG}->{$arg};

}

sub host {
    my $self = shift;

    return $self->{Host};

}

sub port {
    my $self = shift;

    return $self->{Port};

}
    
# ---------------------------------------------------------------------
# Public methods
# ---------------------------------------------------------------------

sub send_data {
    my ($kernel, $self, $frame) = @_[KERNEL, OBJECT, ARG0];

    if (defined($self->{Wheel})) {

        $self->{Wheel}->put($frame);

    }

}

# ---------------------------------------------------------------------
# Public methods, these should be overridden, as needed
# ---------------------------------------------------------------------

sub log {
    my ($self, $kernel, $level, @args) = @_;

    print "$level: @args\n";

}

sub handle_shutdown {
    my ($self, $kernel) = @_;

}

sub handle_connection {
    my ($kernel, $self) = @_[KERNEL, OBJECT];

}

sub handle_connected {
    my ($kernel, $self, $frame) = @_[KERNEL, OBJECT, ARG0];

}

sub handle_message {
    my ($kernel, $self, $frame) = @_[KERNEL, OBJECT, ARG0];

}

sub handle_receipt {
    my ($kernel, $self, $frame) = @_[KERNEL, OBJECT, ARG0];

}

sub handle_error {
    my ($kernel, $self, $frame) = @_[KERNEL, OBJECT, ARG0];

}

sub gather_data {
    my ($kernel, $self, $frame) = @_[KERNEL, OBJECT, ARG0];

}

sub connection_down {
    my ($kernel, $self) = @_[KERNEL, OBJECT];

}

1;


__END__

=head1 NAME

POE::Component::Client::Stomp - Perl extension for the POE Environment

=head1 SYNOPSIS

This module is a class used to create clients that need to access a 
message server that communicates with the STOMP protocol. Your program could 
look as follows:

 package Client;

 use POE;
 use base qw(POE::Component::Client::Stomp);

 use strict;
 use warnings;

 sub handle_connection {
    my ($kernel, $self) = @_[KERNEL, OBJECT];
 
    my $nframe = $self->stomp->connect({login => 'testing', 
                                        passcode => 'testing'});
    $kernel->yield('send_data' => $nframe);

 }

 sub handle_connected {
    my ($kernel, $self, $frame) = @_[KERNEL, OBJECT, ARG0];

    my $nframe = $self->stomp->subscribe({destination => $self->config('Queue'), 
                                          ack => 'client'});
    $kernel->yield('send_data' => $nframe);

 }
 
 sub handle_message {
    my ($kernel, $self, $frame) = @_[KERNEL, OBJECT, ARG0];

    my $message_id = $frame->headers->{'message-id'};
    my $nframe = $self->stomp->ack({'message-id' => $message_id});
    $kernel->yield('send_data' => $nframe);

 }

 package main;

 use POE;
 use strict;

    Client->spawn(
        Alias => 'testing',
        Queue => '/queue/testing',
    );

    $poe_kernel->run();

    exit 0;


=head1 DESCRIPTION

This module handles the nitty-gritty details of setting up the communications 
channel to a message queue server. You will need to sub-class this module
with your own for it to be usefull.

An attempt to maintain that channel will be made when/if that server should 
happen to disappear off the network. There is nothing more unpleasent then 
having to go around to dozens of servers and restarting processes.

When messages are received, specific events are generated. Those events are 
based on the message type. If you are interested in those events you should 
override the default behaviour for those events. The default behaviour is to 
do nothing.

=head1 METHODS

=over 4

=item spawn

This method initializes the class and starts a session to handle the 
communications channel. The only parameters that having meaning are:

=over 4

 Alias           - The session alias, defaults to 'stomp-client'
 RemoteAddress   - The servers hostname, defaults to 'localhost'
 RemotePort      - The servers port, defaults to '61613'
 RetryReconnect  - Wither to attempt reconnections after they run out
 EnableKeepAlive - For those pesky firewalls, defaults to false

=back

All other parameters are stored within an internal config.

=item send_data

You use this event to send Stomp frames to the server. 

=over 4

=item Example

 $kernel->yield('send_data', $frame);

=back

=item handle_connection

This event is signaled and the corresponding method is called upon initial 
connection to the message server. For the most part you should send a 
"connect" frame to the server.

=over 4

=item Example

 sub handle_connection {
     my ($kernel, $self) = @_[KERNEL,$OBJECT];
 
    my $nframe = $self->stomp->connect({login => 'testing', 
                                        passcode => 'testing'});
    $kernel->yield('send_data' => $nframe);
     
 }

=back

=item handled_connected

This event and corresponing method is called when a "CONNECT" frame is 
received from the server. This means the server will allow you to start
generating/processing frames.

=over 4

=item Example

 sub handle_connected {
     my ($kernel, $self, $frame) = @_[KERNEL,$OBJECT,ARG0];
 
     my $nframe = $self->stomp->subscribe({destination => $self->config('Queue'), 
                                           ack => 'client'});
     $kernel->yield('send_data' => $nframe);
     
 }

This example shows you how to subscribe to a particular queue. The queue name
was passed as a parameter to spawn() so it is available in the $self->{CONFIG}
hash.

=back

=item handle_message

This event and corresponding method is used to process "MESSAGE" frames. 

=over 4

=item Example

 sub handle_message {
     my ($kernel, $self, $frame) = @_[KERNEL,$OBJECT,ARG0];
 
     my $message_id = $frame->headers->{'message-id'};
     my $nframe = $self->stomp->ack({'message-id' => $message_id});
     $kernel->yield('send_data' => $nframe);
     
 }

This example really doesn't do much other then "ack" the messages that are
received. 

=back

=item handle_receipt

This event and corresponding method is used to process "RECEIPT" frames. 

=over 4

=item Example

 sub handle_receipt {
     my ($kernel, $self, $frame) = @_[KERNEL,$OBJECT,ARG0];
 
     my $receipt = $frame->headers->{receipt};
     
 }

This example really doesn't do much, and you really don't need to worry about
receipts unless you ask for one when you send a frame to the server. So this 
method could be safely left with the default.

=back

=item handle_error

This event and corresponding method is used to process "ERROR" frames. 

=over 4

=item Example

 sub handle_error {
     my ($kernel, $self, $frame) = @_[KERNEL,$OBJECT,ARG0];
 
 }

This example really doesn't do much. Error handling is pretty much what the
process needs to do when something unexpected happens.

=back

=item gather_data

This event and corresponding method is used to "gather data". How that is done
is up to your program. But usually a "send_data" event is generated.

=over 4

=item Example

 sub gather_data {
     my ($kernel, $self) = @_[KERNEL,$OBJECT];
 
     # doing something here

     $kernel->yield('send_data' => $frame);

 }

=back

=item connection_down

This event and corresponding method is a hook to allow you to be notified if 
the connection to the server is currently down. By default it does nothing. 
But it would be usefull to notify "gather_data" to temporaily stop doing 
whatever it is currently doing.

=over 4

=item Example

 sub connection_down {
    my ($kernel, $self) = @_[KERNEL,OBJECT];

    # do something here

 }

=back

=item log

This method is used internally to send a log message to stdout. It can be 
overridden to hook into your perferred logging module. This module currently
uses the following levels internally: 'warn', 'error', 'debug'

=over 4

=item Example

 sub log {
     my ($self, $kernel, $level, @args) = @_;

     if ($level eq 'error') {

         $kernel->post('logger' => error => @args);

     } elsif ($level eq 'warn') {

         $kernel->post('logger' => warn => @args);

    }

 }

=back

=item handle_shutdown

This method is a hook and should be overidden to do "shutdown" stuff. By
default it does nothing.

=over 4

=item Example

 sub handle_shutdown {
    my ($self, $kernel) = @_;

    # do something here

 }

=back

=back

=head1 ACCESSORS

=over 4

=item stomp

This returns an object to the interal POE::Component::Client::Stomp::Utils 
object. This is very useful for creating Stomp frames.

=over 4

=item Example

 $frame = $self->stomp->connect({login => 'testing', 
                                 passcode => 'testing'});
 $kernel->yield('send_data' => $frame);

=back

=item config

This accessor is used to return items from the internal config. The config is
loaded from the parameters that were used when spawn() was called.

=over 4

=item Example

 $logger = $self->config('Logger');

=back

=back

=head1 SEE ALSO

 Net::Stomp::Frame
 POE::Filter::Stomp
 POE::Component::MessageQueue
 POE::Compoment::Client::Stomp::Utils;

 For information on the Stomp protocol: http://stomp.codehaus.org/Protocol

=head1 AUTHOR

Kevin L. Esteb, E<lt>kesteb@wsipc.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by Kevin L. Esteb

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut
